public/scripts/customInput.js

```diff
@@ -0,0 +1,191 @@
+'use strict'
+
+const {
+  socket,
+  addListItem,
+  addListItemAsPenultimateChild
+} = require('./shared')
+
+const body = document.querySelector('body')
+const wordCounter = document.querySelector('#wordCounter')
+const errorSound = document.querySelector('#errorSound')
+
+const VALID_CHARS_REGEXP = /^[A-Z]$/
+
+// setup the list within the customInput tag
+const inputElement = document.querySelector('#customInput')
+const words = document.createElement('ul')
+words.classList.add('words')
+inputElement.appendChild(words)
+
+let currentWord
+const wordSet = new Set()
+let RANDOM_WORDS
+let WORD_COUNT
+const FRACTION_REQUIRED_TO_BE_RANDOM = 0.50
+
+function keydownListener(event) {
+  const key = event.key.toUpperCase()
+  if (VALID_CHARS_REGEXP.test(key)) {
+    handleValidChars(key)
+  } else {
+    switch (key) {
+      case ' ':
+        handleSpace()
+        break
+      case 'BACKSPACE':
+        handleBackspace()
+        break
+      case 'ENTER':
+        handleEnter()
+        break
+      default:
+        handleBadKeystrokes()
+    }
+  }
+  updateWordCounter()
+}
+
+function resetWords() {
+  words.innerHTML = ''
+  addListItem(words, ['word', 'cursor'], ' ')
+  currentWord = addListItemAsPenultimateChild(words, ['word'])
+  updateWordCounter()
+}
+
+function updateWordCounter() {
+  let numWords = words.querySelectorAll('.word:not(.cursor)').length
+  if (currentWord.textContent === '') {
+    --numWords
+  }
+  wordCounter.textContent = `${numWords}/${WORD_COUNT}`
+  if (numWords !== WORD_COUNT) {
+    wordCounter.classList.add('error')
+  } else {
+    wordCounter.classList.remove('error')
+  }
+}
+
+function classifyWord() {
+  const currentText = currentWord.textContent
+  currentWord.className = 'word'
+  if (wordSet.has(currentText)) {
+    currentWord.classList.add('error')
+  } else {
+    if (RANDOM_WORDS.includes(currentText)) {
+      currentWord.classList.add('keyword')
+      toggleMarkForWord(currentText)
+    }
+  }
+}
+
+function toggleMarkForWord(word) {
+  const requiredWords = document.querySelectorAll('.requiredWord')
+  for (let i = 0; i < requiredWords.length; ++i) {
+    if (requiredWords[i].textContent === word) {
+      requiredWords[i].classList.toggle('marked')
+    }
+  }
+}
+
+function handleValidChars(key) {
+  event.preventDefault()
+  currentWord.textContent += key
+  classifyWord()
+}
+
+function handleSpace() {
+  if (currentWord.textContent === '') {
+    return // do nothing if no word to add
+  }
+  wordSet.add(currentWord.textContent)
+  currentWord = addListItemAsPenultimateChild(words, ['word'])
```

Comment:
Yeah, I was getting a bit overwhelmed with the possible features and I want to keep moving forward. One, so I don't block you guys from working on top of the fundamental parts I built, and Two, so we can have smaller PRs more frequently, which is the easier and better way of doing things I think